File: 567.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s2.size() < s1.size()) return false;
        vector<int> alpha(26, 0);
        for (int i = 0; i < s1.size(); ++i) {
            alpha[s1[i] - 'a']++;
        }
        for (int i = 0; i < s1.size(); ++i) {
            alpha[s2[i] - 'a']--;
        }
        if (check(alpha)) return true;
        for (int i = 1; i + s1.size() <= s2.size(); ++i) {
            alpha[s2[i-1] - 'a']++;
            alpha[s2[i + s1.size() - 1] - 'a']--;
            if (check(alpha)) return true;
        }
        return false;
    }
    bool check(vector<int> &alpha) {
        for (int i = 0; i < alpha.size(); ++i) {
            if (alpha[i] > 0)
                return false;
        }
        return true;
    }
};

File: 567.go
package leetcode

func checkInclusion(s1 string, s2 string) bool {
	if (len(s2) < len(s1)) {
		return false
	}
	alpha := make([]int, 26)
	check := func (alpha []int) bool {
		for i := 0; i < 26; i++ {
			if (alpha[i] > 0) {
				return false
			}
		}
		return true
	}
	for i := 0; i < len(s1); i++ {
		alpha[s1[i] - 'a']++
	}	
	for i := 0; i < len(s1); i++ {
		alpha[s2[i] - 'a']--
	}	
	if (check(alpha)) {
		return true
	}
	for i := 1; i + len(s1) <= len(s2); i++ {
		alpha[s2[i - 1] - 'a']++
		alpha[s2[i + len(s1) - 1] - 'a']--
		if (check(alpha)) {
			return true
		}
	}
	return false
}	